### ISSUE TITLE
$invoice->coupon() returns faulty value when tax is applied

### ISSUE THREAD

**Issue by tobiasvielmetter** [laravel/cashier#374]
I just realised when I apply tax and a percentage coupon, the method $invoice->coupon(), which shows the amount off, is calculated wrongly, as it subtracts from the invoice total after taxes.
the method looks like this:
`public function discount()
    {
        return $this->formatAmount($this->invoice->subtotal - $this->invoice->total);
    }`

this is my dirty workaround:
`public function discount()
    {
        return $this->formatAmount($this->invoice->subtotal - ($this->tax?($this->invoice->total-$this->tax):$this->invoice->total));
    }`

I am happy to make pull request for this – do let me know if this is ok and I will.

**Comment by Mr-Anonymous** [comment]
@hellohereiam You are exactly right. I have the exact same issue where I charge 10% GST for my users. And the invoice shows like this:
```
Subscription (2) March 7, 2017 - March 8, 2017 $2.00
COUPON50 (50% Off) -$0.90
GST (10%) $0.10
```
This is incorrect. Your suggested method of `subtotal - (total - tax)` is the right method to calculate the discount amount when tax is applied. 

Would they add this to the core since it is mathematical wrong as it doesn't equate scenario when Tax is included in the invoice.

**Comment by Mr-Anonymous** [comment]
Is there a way to add this correctly directly in the `receipt.blade.php` file or the `Invoice Controller` without needing the edit the source file `cashier/src/Invoice.php`? How can I calculate this from Controller or Blade?

**Comment by Mr-Anonymous** [comment]
Ok, after a few trial & error, this is how I have added the discount calculation directly into my `receipt.blade.php` file.
```
{{ Laravel\Cashier\Cashier::formatAmount($invoice->subtotal - ($invoice->tax?($invoice->total-$invoice->tax):$invoice->total)) }}
```
The above seems to be working. If someone thinks this could cause issue due to having this in blade file directly, please do let me know since I will be adding this to my production site in 2 days.

**Comment by hellohereiam** [comment]
@Mr-Anonymous this looks good to me. I have done it in a similar fashion, except that I have done the calculation int he controller (which seemed the more appropriate position, but in the end its the same).
You should also have a look at #384 as it is a similar issue you will be running into, once your customers swap plans (the proration invoice comes without tax unless you add it yourself).

**Comment by Mr-Anonymous** [comment]
Thank you @hellohereiam I appreciate your feedback. That certainly helps. 

I wasnt aware of the proration issue with tax and I never noticed it when testing it. I added the tax rate directly from Stripe dashboard and not via Cashier charge and I wonder if the same issue happens there. I will take another look at it. Thank you for pointing it out.

Cheers,
Neel.

**Comment by hellohereiam** [comment]
@Mr-Anonymous probably the issue is not there if you added the tax rate in stripe. I have variable tax rates, so thats not an option for me.

**Comment by Mr-Anonymous** [comment]
@hellohereiam  I think you are right. Since the tax is handled by Stripe in my case, prorationed seems to work. I have 1 tax rate and 1 subscription plan so Cashier only handles the subscription increment, decrement, activation and cancellation. So a simple straight forward set-up in my case. 

Thanks for your comments. Much obliged.

Cheers,
Neel.

**Comment by itsDamien** [comment]
Hey guys !
Fixed in #481

**Comment by mkarnicki** [comment]
@tobiasvielmetter Please close this issue.
